Broken (read-only) database connections not getting removed from connection pool, keycloak claims to be healthy.

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

We have a master/standby setup of our postgres database (standard zalando/postgres-operator).

We had a "hiccup" in our k8s cluster, which resulted in  a failover - master and standby got switched. The peculiar thing about this switch was that the former master survived the hiccup, only the former standby got restarted and promoted to master.

This resulted in all connections to the database in the Keycloak pool still being active, but now pointing to the standby database, meaning they are read-only.

From that moment on, Keycloak stopped working properly. Every login failed with this exception:

```
2023-10-31 18:47:47.096	
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:197)
2023-10-31 18:47:47.096	
	at io.agroal.pool.wrapper.PreparedStatementWrapper.executeUpdate(PreparedStatementWrapper.java:88)
2023-10-31 18:47:47.096	
	at com.sun.proxy.$Proxy86.executeUpdate(Unknown Source)
2023-10-31 18:47:47.096	
	at org.postgresql.ds.PGPooledConnection$StatementHandler.invoke(PGPooledConnection.java:441)
2023-10-31 18:47:47.096	
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2023-10-31 18:47:47.096	
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-10-31 18:47:47.096	
	at jdk.internal.reflect.GeneratedMethodAccessor181.invoke(Unknown Source)
2023-10-31 18:47:47.096	
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:152)
2023-10-31 18:47:47.096	
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:190)
2023-10-31 18:47:47.096	
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:413)
2023-10-31 18:47:47.096	
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:496)
2023-10-31 18:47:47.096	
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:356)
2023-10-31 18:47:47.096	
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2366)
2023-10-31 18:47:47.096	
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2676)

2023-10-31 18:47:47.096	
Caused by: org.postgresql.util.PSQLException: ERROR: cannot execute INSERT in a read-only transaction
```

This could only be resolved by manually restarting the keycloak server to force a reconnection to the new master database.


### Version

quay.io/keycloak/keycloak:20.0.5 installed via helm chart

### Expected behavior

Database connection that cause exceptions should be removed from the pool and not be reused. No manual intervention should be necessary.

### Actual behavior

Keycloak is using the non-functional database connections ad inifinitum and doesn't care about exceptions. Health check is reporting "healthy", though the database connection is read only.

### How to Reproduce?

If you do not want to setup the whole HA feature, start Keycloak with a local Postgres and set it to read only: 
`alter system set default_transaction_read_only=on;`

### Anything else?

We have a lot of Spring application in our cluster. They are using the Hikari connection pool. This pool can be configured so that database connections which run into an exception are always and immediately removed from the connection pool. This eventually auto-heals the application, as the read only connection get dropped and replaced by fresh connections to the master database.

As Keycloak is using the Agroal pool, I found this is the documenation:

> exceptionSorter(ExceptionSorter) - This extension point tells Agroal which exceptions indicate that the connection is in an invalid state and should be removed from the pool. Agroal provides a few implementations of the ExceptionSorter interface for selected databases. By default, no exception is considered fatal.

It would be nice if you could activate this feature.

--- a/model/jpa/src/main/java/org/keycloak/connections/jpa/DefaultJpaConnectionProviderFactory.java
+++ b/model/jpa/src/main/java/org/keycloak/connections/jpa/DefaultJpaConnectionProviderFactory.java
@@ -178,9 +178,7 @@ public class DefaultJpaConnectionProviderFactory implements JpaConnectionProvide
                         } else {
                             String url = config.get("url");
                             String driver = config.get("driver");
-                            if (driver.equals("org.h2.Driver")) {
-                                url = addH2NonKeywords(url);
-                            }
+                            url = augmentJdbcUrl(driver, url);
                             properties.put(AvailableSettings.JAKARTA_JDBC_URL, url);
                             properties.put(AvailableSettings.JAKARTA_JDBC_DRIVER, driver);
 
@@ -373,15 +371,23 @@ public class DefaultJpaConnectionProviderFactory implements JpaConnectionProvide
             } else {
                 String url = config.get("url");
                 String driver = config.get("driver");
-                if (driver.equals("org.h2.Driver")) {
-                    url = addH2NonKeywords(url);
-                }
+                url = augmentJdbcUrl(driver, url);
                 Class.forName(driver);
                 return DriverManager.getConnection(StringPropertyReplacer.replaceProperties(url, System.getProperties()), config.get("user"), config.get("password"));
             }
         } catch (Exception e) {
             throw new RuntimeException("Failed to connect to database", e);
         }
+    }
+
+    private String augmentJdbcUrl(String driver, String url) {
+        if (driver.equals("org.postgresql.xa.PGXADataSource") || driver.equals("org.postgresql.Driver")) {
+            url = addPostgreSQLKeywords(url);
+        }
+        if (driver.equals("org.h2.Driver")) {
+            url = addH2NonKeywords(url);
+        }
+        return url;
     }
 
     @Override
@@ -431,6 +437,24 @@ public class DefaultJpaConnectionProviderFactory implements JpaConnectionProvide
     private String addH2NonKeywords(String jdbcUrl) {
         if (!jdbcUrl.contains("NON_KEYWORDS=")) {
             jdbcUrl = jdbcUrl + ";NON_KEYWORDS=VALUE";
+        }
+        return jdbcUrl;
+    }
+
+    /**
+     * For a PostgreSQL cluster, Keycloak would need to connect to the primary node that is writable.
+     * The `targetServerType` should avoid connecting to a reader instance accidentally during node failover.
+
+     * @return JDBC URL with <code>targetServerType=primary</code> appended if the URL doesn't contain <code>targetServerType=</code> yet
+     */
+    private String addPostgreSQLKeywords(String jdbcUrl) {
+        if (!jdbcUrl.contains("targetServerType=")) {
+            if (jdbcUrl.contains("?")) {
+                jdbcUrl = jdbcUrl + "&";
+            } else {
+                jdbcUrl = jdbcUrl + "?";
+            }
+            jdbcUrl = jdbcUrl + "targetServerType=primary";
         }
         return jdbcUrl;
     }

